### ISSUE TITLE
Request: Problems loading CaffeineCacheSupplier in Waffle-JNA/Waffle-JNA-Jakarta

### ISSUE THREAD

**Issue by FarmGeek4Life** [Waffle/waffle#2522]
I use Waffle-JNA (and now Waffle-JNA-Jakarta) in a plugin I maintain for Jenkins CI; I am able to use the 2.x versions without issue, but if I try to update to the 3.x versions the ServiceLoader.load(CacheSupplier.class) call in waffle.util.cache.Cache always fails (NoSuchElement exception is thrown); for some reason the overall system is ignoring the META-INF/services file that is added in both the waffle-jna dependency and that I have added to my plugin.

My request: Could the default CaffeineCacheSupplier be loaded programmatically as a fallback, probably with a logging message added in? Unless I can figure this out, I am stuck on the 2.x versions.

**Comment by hazendaz** [comment]
Can you provide a stack trace?

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Bryson Gibbons ***@***.***>
Sent: Thursday, September 12, 2024 3:26:34 PM
To: Waffle/waffle ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [Waffle/waffle] Request: Problems loading CaffeineCacheSupplier in Waffle-JNA/Waffle-JNA-Jakarta (Issue #2522)


I use Waffle-JNA (and now Waffle-JNA-Jakarta) in a plugin I maintain for Jenkins CI; I am able to use the 2.x versions without issue, but if I try to update to the 3.x versions the ServiceLoader.load(CacheSupplier.class) call in waffle.utils.cache.Cache always fails (NoSuchElement exception is thrown); for some reason the overall system is ignoring the META-INF/services file that is added in both the waffle-jna dependency and that I have added to my plugin.

My request: Could the default CaffeineCacheSupplier be loaded programmatically as a fallback, probably with a logging message added in? Unless I can figure this out, I am stuck on the 2.x versions.

**Comment by FarmGeek4Life** [comment]
Part of the challenge is I don't think I am getting the full stack trace; at the very least, if there are nested/inner exceptions, they are not being reported. I supposed I could wrap the call in my code with an exception handler and independently log the exception and any nested exceptions.

```
Failed to instantiate Key[type=com.github.farmgeek4life.jenkins.negotiatesso.NegotiateSSO, annotation=[none]]; skipping this component
java.util.NoSuchElementException
	at PluginClassLoader for NegotiateSSO//waffle.util.cache.Cache.newCache(Cache.java:62)
	at PluginClassLoader for NegotiateSSO//waffle.windows.auth.impl.WindowsAuthProviderImpl.<init>(WindowsAuthProviderImpl.java:103)
	at PluginClassLoader for NegotiateSSO//waffle.windows.auth.impl.WindowsAuthProviderImpl.<init>(WindowsAuthProviderImpl.java:93)
	at PluginClassLoader for NegotiateSSO//waffle.servlet.NegotiateSecurityFilter.init(NegotiateSecurityFilter.java:387)
	at PluginClassLoader for NegotiateSSO//com.github.farmgeek4life.jenkins.negotiatesso.NegotiateSSO.startFilter(NegotiateSSO.java:171)
	at PluginClassLoader for NegotiateSSO//com.github.farmgeek4life.jenkins.negotiatesso.NegotiateSSO.start(NegotiateSSO.java:126)
	at PluginClassLoader for NegotiateSSO//com.github.farmgeek4life.jenkins.negotiatesso.NegotiateSSO.<init>(NegotiateSSO.java:108)
	at PluginClassLoader for NegotiateSSO//com.github.farmgeek4life.jenkins.negotiatesso.NegotiateSSO$$FastClassByGuice$$1701a7ef.GUICE$TRAMPOLINE(<generated>)
	at PluginClassLoader for NegotiateSSO//com.github.farmgeek4life.jenkins.negotiatesso.NegotiateSSO$$FastClassByGuice$$1701a7ef.apply(<generated>)
	at com.google.inject.internal.DefaultConstructionProxyFactory$FastClassProxy.newInstance(DefaultConstructionProxyFactory.java:82)
	at com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:114)
	at com.google.inject.internal.ConstructorInjector.access$000(ConstructorInjector.java:33)
	at com.google.inject.internal.ConstructorInjector$1.call(ConstructorInjector.java:98)
	at com.google.inject.internal.ProvisionListenerStackCallback$Provision.provision(ProvisionListenerStackCallback.java:109)
	at hudson.ExtensionFinder$GuiceFinder$SezpozModule.onProvision(ExtensionFinder.java:613)
	at com.google.inject.internal.ProvisionListenerStackCallback$Provision.provision(ProvisionListenerStackCallback.java:117)
	at com.google.inject.internal.ProvisionListenerStackCallback.provision(ProvisionListenerStackCallback.java:66)
	at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:93)
	at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:300)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
Caused: com.google.inject.ProvisionException: Unable to provision, see the following errors:

1) [Guice/ErrorInjectingConstructor]: NoSuchElementException
  at NegotiateSSO.<init>(NegotiateSSO.java:105)

Learn more:


Learn more:
  https://github.com/google/guice/wiki/ERROR_INJECTING_CONSTRUCTOR

1 error

======================
Full classname legend:
======================
NegotiateSSO:           "com.github.farmgeek4life.jenkins.negotiatesso.NegotiateSSO"
========================
End of classname legend:
========================

	at com.google.inject.internal.InternalProvisionException.toProvisionException(InternalProvisionException.java:251)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:43)
	at com.google.inject.internal.SingletonScope$1.get(SingletonScope.java:169)
	at hudson.ExtensionFinder$GuiceFinder$FaultTolerantScope$1.get(ExtensionFinder.java:448)
	at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:45)
	at com.google.inject.internal.InjectorImpl$1.get(InjectorImpl.java:1148)
	at hudson.ExtensionFinder$GuiceFinder._find(ExtensionFinder.java:406)
	at hudson.ExtensionFinder$GuiceFinder.find(ExtensionFinder.java:397)
	at hudson.ClassicPluginStrategy.findComponents(ClassicPluginStrategy.java:353)
	at hudson.ExtensionList.load(ExtensionList.java:384)
	at hudson.ExtensionList.ensureLoaded(ExtensionList.java:320)
	at hudson.ExtensionList.getComponents(ExtensionList.java:184)
	at hudson.DescriptorExtensionList.load(DescriptorExtensionList.java:213)
	at hudson.ExtensionList.ensureLoaded(ExtensionList.java:320)
	at hudson.ExtensionList.iterator(ExtensionList.java:172)
	at PluginClassLoader for xunit//org.jenkinsci.plugins.xunit.AliasInitializer.init(AliasInitializer.java:46)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at hudson.init.TaskMethodFinder.invoke(TaskMethodFinder.java:109)
	at hudson.init.TaskMethodFinder$TaskImpl.run(TaskMethodFinder.java:185)
	at org.jvnet.hudson.reactor.Reactor.runTask(Reactor.java:305)
	at jenkins.model.Jenkins$5.runTask(Jenkins.java:1195)
	at org.jvnet.hudson.reactor.Reactor$2.run(Reactor.java:221)
	at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:120)
	at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:68)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
```

**Comment by FarmGeek4Life** [comment]
And, decided to do more testing, even going as far as trying the ServiceLoader call in my own code so I can track any usages: there are no suppressed exceptions, and ServiceLoader is not returning any CacheSupplier classes, even when I provide my own copy of META-INF/services/waffle.util.cache.CacheSupplier.

And... I finally found something relevant for me to work around the problem: https://github.com/jenkinsci/saml-plugin/blob/main/src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java
I need to temporarily modify the thread's context class loader so that it is operating within the context of waffle-jna...

**Comment by hazendaz** [comment]
Have you been able to get that working based on that information?  Seems like a FAQ opportunity on this.  If so, could you add some documentation on this to FAQ.  General area would be somewhere around https://github.com/Waffle/waffle/blob/3090acf5b80f885171960a511115e63f5fc854aa/Docs/faq/CustomCache.md (possibly the file calling that then information on how to address in case you have.

There could be an opportunity to add as well programmatic way you mention.  I believe slf4j/logback did that recently due to possibly similar issues.  So if you search their repos you should be able to find what they did.

If you could contribute that would be great.  I'm wanting to do a release of this and trying to determine if I need to pick up something here.  If you think you could contribute something to solve this, I can wait on a release.  If it looks like just a documentation update will suffice, I'll go ahead and continue with release need I need to push.  I'm not in any big rush so I can wait a few weeks here.

**Comment by hazendaz** [comment]
Released as 3.5.0, will be available in central in hour or so.
